Bottom: 08c12f7aaf55eff2895ccaf1e1edcc597cbe3054
Top:    4e37d1a24ca55fc17f8a096b813467c85eec8c90
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-12-13 01:15:17 +0000

hid/ps: Add support for rendering "fullpoly" polygons

Affects-bug: lp-1089532


---

diff --git a/src/hid/ps/ps.c b/src/hid/ps/ps.c
index 9a95854..9080787 100644
--- a/src/hid/ps/ps.c
+++ b/src/hid/ps/ps.c
@@ -1244,7 +1244,7 @@ ps_fill_polygon (hidGC gc, int n_coords, Coord *x, Coord *y)
 }
 
 static void
-ps_draw_pcb_polygon (hidGC gc, PolygonType * poly, const BoxType * clip_box)
+fill_polyarea (hidGC gc, POLYAREA * pa, const BoxType * clip_box)
 {
   /* Ignore clip_box, just draw everything */
 
@@ -1254,7 +1254,7 @@ ps_draw_pcb_polygon (hidGC gc, PolygonType * poly, const BoxType * clip_box)
 
   use_gc (gc);
 
-  pl = poly->Clipped->contours;
+  pl = pa->contours;
 
   do
     {
@@ -1273,6 +1273,19 @@ ps_draw_pcb_polygon (hidGC gc, PolygonType * poly, const BoxType * clip_box)
 }
 
 static void
+ps_draw_pcb_polygon (hidGC gc, PolygonType * poly, const BoxType * clip_box)
+{
+  fill_polyarea (gc, poly->Clipped, clip_box);
+  if (TEST_FLAG (FULLPOLYFLAG, poly))
+    {
+      POLYAREA *pa;
+
+      for (pa = poly->Clipped->f; pa != poly->Clipped; pa = pa->f)
+        fill_polyarea (gc, pa, clip_box);
+    }
+}
+
+static void
 ps_fill_rect (hidGC gc, Coord x1, Coord y1, Coord x2, Coord y2)
 {
   use_gc (gc);
